### ISSUE TITLE
Random crash using Timestretch in Windows 64bit

### ISSUE THREAD

**Issue by jimsimonz** [superpoweredSDK/Low-Latency-Android-iOS-Linux-Windows-tvOS-macOS-Interactive-Audio-Platform#561]
Getting a very intermittent crash in timestretch on Windows. Probably one in 100 plays.

There's no pattern to when it happens other than it always goes bang in the same place in  SuperpoweredFFTReal and happens on the very first process call where 2048 frames are requested.

Callstack:
 	!SuperpoweredFFTReal(float *,float *,int,bool)	Unknown
 	!SuperpoweredFFTComplex(float *,float *,int,bool)	Unknown
 	!SuperpoweredPolarFFT(float *,float *,int,bool,float)	Unknown
 	SuperpoweredFrequencyDomain::timeDomainToFrequencyDomain(float *,float *,float *,float *,float,bool,int)	Unknown
 	SuperpoweredTimeStretching::process(struct SuperpoweredAudiobufferlistElement *,class SuperpoweredAudiopointerList *)	Unknown

The exception is 0xC0000005: Access violation reading location 0x00000253197D0780.
This address is nowhere near the supplied input buffer address. Also why in the examples does the input buffer size get +64 added to it. I do add this and tried adding larger padding but no difference to the crash.

Tried many things to work around/figure it out and can't from outside the library.

**Comment by gaborszanto** [comment]
One in 100 plays is definitely too much. Is this happening with one of the example projects or with your own code? If it happens with your code, can you change the player example project (with setTempo, so it uses time stretching) to run that many times to see if it crashes?

**Comment by jimsimonz** [comment]
It is with my own code though I have tested with other time stretch SDKs with the same surrounding code and there's no crash so I'm assuming the problem's in the library.

I'll try a test rig as soon as I can with your sample code I doubt it does the same thing as my code is running a small subset of just the timestretch process part with input and output buffers.

My code is highly multithreaded (however, each instance of superpowered is only accessed by the one thread)

**Comment by jimsimonz** [comment]
One additional question. Are there any resource limitations with superpowered timestretcher?

The easiest way to repro the crash is to run a lot of them simultaneously. So for example I pitch shift a drum sound that's 5 seconds long, then do a drum roll which ends up with a polyphony of maybe 20. This crashes it as if it reaches an allocation limit internally.

**Comment by gaborszanto** [comment]
There are no resource limitations, although i never run more than 16 instance simultaneously (but i do 16 all the time in my DJ app and it works great even on the many years old iPad 2).
How far is the access violation address from the input buffer address? How do you allocate the input buffer?

**Comment by jimsimonz** [comment]
Thanks for the response.

I use the getbuffer call as in the sample code for the input buffers. My code is pretty much a copy of the sample but wrapped with an intermidiate buffer and using the requested sample count from superpowered.


// Create an input buffer for the time stretcher.
	SuperpoweredAudiobufferlistElement InputBuffer;
	InputBuffer.samplePosition = 0;
	InputBuffer.startSample = 0;
	InputBuffer.samplesUsed = 0;
	InputBuffer.endSample = NumRawPCMFrames; // <-- Important!

	InputBuffer.buffers[0] = InputBuffer.buffers[1] = InputBuffer.buffers[2] = InputBuffer.buffers[3] = NULL;

	int32 AllocSize = NumRawPCMFrames * Wave->NumChannels * sizeof(float);
	// from the sample, no idea why you have to add 64?
	AllocSize = AllocSize + 128;
	InputBuffer.buffers[0] = SuperpoweredAudiobufferPool::getBuffer(AllocSize);
	FMemory::Memzero(InputBuffer.buffers[0], AllocSize);
	// assumes stereo frames so half num for mono?
	SuperpoweredShortIntToFloat(SourceBuffer, (float*)InputBuffer.buffers[0], (NumRawPCMFrames *  NumChannels) / 2);
	TimeStretcher->process(&InputBuffer, OutputBuffers);

**Comment by gaborszanto** [comment]
Is it possible that SuperpoweredAudiobufferPool::getBuffer returned with a NULL?
In the call stack, which parameter of SuperpoweredFFTReal is near 0x00000253197D0780?

**Comment by jimsimonz** [comment]
<<Is it possible that SuperpoweredAudiobufferPool::getBuffer returned with a NULL?
I'll add a double check though I'm surprised it got that far down the call stack if so.
<<In the call stack, which parameter of SuperpoweredFFTReal is near 0x00000253197D0780?
The callstack doesn't have that much detail as it's a minidump. I'll try and catch in in a debug version.

I think the way to proceed from here is that I'll create test code that can crash it 100 percent in isolation of the larger app and send this to you for debugging. I don't have time right now to work on the test code though.

**Comment by gaborszanto** [comment]
thank you, looking forward to it

**Comment by jimsimonz** [comment]
Just to update the status on this issue, it's not a NULL input buffer.

**Comment by gaborszanto** [comment]
I'm looking forward to your debug mode results.
Meanwhile just a thought: is it possible that when the crash happens you're already in the middle of removing the time stretcher instance? Many other time stretchers have synchronization (such as mutexes) inside, however Superpowered doesn't any (for highest performance).

**Comment by jimsimonz** [comment]
I tried wrapping the whole thing in a critical section initially just as a test and It didn't help. It is accessed from one worker thread anyway though. The behaviour does look like either multithreaded or an uninitialised due to the random nature of it.

**Comment by gaborszanto** [comment]
Are you still trying to get in the call stack, which parameter of SuperpoweredFFTReal is near 0x00000253197D0780?

**Comment by jimsimonz** [comment]
I can't easily get at the params and can't link to the debug libs without a
mission so leaving this for now and using Rubberband. I can then fix that
code if there's an issue. It's a pity as Superpowered sounds a little
better at least on default settings.

FYI Sometimes the crash is on the very first playback. Either it's an
uninitialised or an over/underrun.

**Comment by gaborszanto** [comment]
Literally thousands of apps are using Superpowered, including many high profile DJ apps heavily time-stretching all the time. There must be a solution.
We'll release a new version soon. Please test that if you can, because it will have some internal changes.

**Comment by jimsimonz** [comment]
Thanks I'll give it a try.

**Comment by tabahkanlahhati** [comment]
On issue

**Comment by gaborszanto** [comment]
The update is out.

**Comment by carmethene** [comment]
@gaborszanto @jimsimonz we have hit this exact crash in the FFT code too.  In our case it came from within the player's time stretching code.  We think it has something to do with running the analyzer in parallel to using master tempo on the audio player, because doing either in isolation doesn't trigger the crash.

That said, we haven't seen it in a while.  The most significant change I have made since I last saw the crash was to write the analyzer management to run on frame-sized chunks rather than the entire track in one go.

**Comment by jimsimonz** [comment]
Thanks for the update, has windows static library changed? https://github.com/superpoweredSDK/Low-Latency-Android-iOS-Linux-Windows-tvOS-macOS-Interactive-Audio-Platform/blob/master/Superpowered/Windows/SuperpoweredWin141_Release_MT_x64.lib or is it just rebuilt?

**Comment by jimsimonz** [comment]
@carmethene Thanks for the info, good to see it's not just my codebase. I changed the padding to 128 and it still has issues occasionally.

**Comment by gaborszanto** [comment]
static lib changed

**Comment by jimsimonz** [comment]
Hi, it still crashes in the same place with the new build.

**Comment by gaborszanto** [comment]
It would really help if you could debug with the debug builds.

**Comment by gaborszanto** [comment]
Is there any activity on this? I pushed out two updates since...

**Comment by jimsimonz** [comment]
I reported it as not fixed with 1.30, I'll have another go with 1.31 has there been an attempt to fix it in 1.31?

**Comment by gaborszanto** [comment]
I never attempted to fix this issue, because i'm not able to reproduce it and the other user commenting had his issue resolved last year. No other users reported any similar problem.

Significant parts of the library has been changed and i thought that maybe it will have an impact on your issue.

**Comment by gaborszanto** [comment]
I think the right way to go forward with this is as you mentioned: "I think the way to proceed from here is that I'll create test code that can crash it 100 percent in isolation of the larger app and send this to you for debugging."

**Comment by jimsimonz** [comment]
Yes that would be ideal but I'm now using another SDK (at least for Windows) and don't want to spend more time on Superpowered right now. 

If I have time I may return to this and create a test rig. 

I retested with your latest commit and it still crashes. 
It can take a couple of hours of testing to break it. I'm certain this is a bug in Superpowered given the stability without this SDK. It looks like under the hood you have an async allocator? So maybe this has a window where it's not thread safe? 

Anyways for the moment I'm leaving it alone.

**Comment by gaborszanto** [comment]
Yes, it has a non-conventional allocator to allocate memory on the audio processing thread without any blocking, however that part is completely thread-safe and we received no bugs or crashes related to the allocator. It allocates memory in 16 MB chunks btw. I wonder if you've any memory leaks so it can not keep up? Did you check your overall memory usage?

**Comment by jimsimonz** [comment]
There's no memory leaks and sometimes it can crash on the first use. It's random so usually takes a lot of testing to break it. I use another sdk in an identical way with no issues. Perhaps postpone the issue? I can reopen if I come back to it.

**Comment by gaborszanto** [comment]
I agree.
